apps/sms/js/activity_handler.js

```diff
@@ -300,62 +305,51 @@ var ActivityHandler = {
-  // Deliver the user to the correct view
-  // based on the params provided in the
-  // "message" object.
-  //
-  toView: function ah_toView(message) {
-    /**
-     *  "message" is either a message object that belongs
-     *  to a thread, or a message object from the system.
-     *
-     *
-     *  message {
-     *    number: A string phone number to pre-populate
-     *            the recipients list with.
-     *
-     *    body: An optional body to preset the compose
-     *           input with.
-     *
-     *    contact: An optional "contact" object
-     *
-     *    threadId: An option threadId corresponding
-     *              to a new or existing thread.
-     *
-     *  }
-     */
 
+  /**
+   * Delivers the user to the correct view based on the params provided in the
+   * "message" parameter.
+   * @param {{number: string, body: string, contact: MozContact,
+   * threadId: number}} message It's either a message object that belongs to a
+   * thread, or a message object from the system. "number" is a string phone
+   * number to pre-populate the recipients list with, "body" is an optional body
+   * to preset the compose input with, "contact" is an optional MozContact
+   * instance, "threadId" is an optional threadId corresponding to a new or
+   * existing thread.
+   * @param {Boolean} focusComposer Indicates whether we need to focus composer
+   * when we navigate to Thread panel.
+   */
+  toView: function ah_toView(message, focusComposer) {
     if (!message) {
       return;
     }
 
     this.isLocked = false;
-    var threadId = message.threadId ? message.threadId : null;
-    var body = message.body || '';
-    var number = message.number ? message.number : '';
-    var contact = message.contact ? message.contact : null;
 
-    var showAction = function act_action() {
+    var navigateToView = function act_navigateToView() {
       // If we only have a body, just trigger a new message.
-      if (!threadId) {
-        ActivityHandler.triggerNewMessage(body, number, contact);
+      if (!message.threadId) {
+        ActivityHandler.triggerNewMessage(
+          message.body, message.number, message.contact
+        );
         return;
       }
 
-      Navigation.toPanel('thread', { id: threadId }).then(Compose.focus);
+      Navigation.toPanel(
+        'thread', { id: message.threadId, focusComposer: focusComposer }
+      );
     };
 
     navigator.mozL10n.once(function waitLocalized() {
       if (!document.hidden) {
         // Case of calling from Notification
-        showAction();
+        navigateToView();
         return;
       }
 
-      document.addEventListener('visibilitychange',
-        function waitVisibility() {
-          document.removeEventListener('visibilitychange', waitVisibility);
-          showAction();
+      document.addEventListener('visibilitychange', function waitVisibility() {
+        document.removeEventListener('visibilitychange', waitVisibility);
+        navigateToView();
```

Comment:
I don't really remember why we wait for a visibility change... but let's not change this in a blocker.